ext/mbstring/mbstring.c

```diff
@@ -2842,16 +2836,163 @@ static const mbfl_encoding **duplicate_elist(const mbfl_encoding **elist, size_t
+	 * codepoints we are likely to see in practice, while only covering
+	 * a small minority of the entire Unicode encoding space. Why?
+	 * Well, if the test string happens to be valid in an incorrect
+	 * candidate encoding, the bogus codepoints which it decodes to will
+	 * be more or less random. By treating the majority of codepoints as
+	 * 'rare', we ensure that in almost all such cases, the bogus
+	 * codepoints will include plenty of 'rares', thus giving the
+	 * incorrect candidate encoding lots of demerits. See
+	 * common_codepoints.txt for the actual list used.
+	 *
+	 * So, why give extra demerits for ASCII punctuation characters? It's
+	 * because there are some text encodings, like UTF-7, HZ, and ISO-2022,
+	 * which deliberately only use bytes in the ASCII range. When
+	 * misinterpreted as ASCII/UTF-8, strings in these encodings will
+	 * have an unusually high number of ASCII punctuation characters.
+	 * So giving extra demerits for such characters will improve
+	 * detection accuracy for UTF-7 and similar encodings.
+	 *
+	 * Finally, why 1 demerit for all other characters? That penalizes
+	 * long strings, meaning we will tend to choose a candidate encoding
+	 * in which the test string decodes to a smaller number of
+	 * codepoints. That prevents single-byte encodings in which almost
+	 * every possible input byte decodes to a 'common' codepoint from
+	 * being favored too much. */
+	if (w > 0xFFFF) {
+		return 40;
+	} else if (w >= 0x21 && w <= 0x2F) {
+		return 6;
+	} else if ((rare_codepoint_bitvec[w >> 5] >> (w & 0x1F)) & 1) {
+		return 30;
+	} else {
+		return 1;
+	}
+	return 0;
+}
+
+/* When doing 'strict' detection, any string which is invalid in the candidate encoding
+ * is rejected. With non-strict detection, we just continue, but apply demerits for
+ * each invalid byte sequence */
+static const mbfl_encoding* mb_guess_encoding(unsigned char *in, size_t in_len, const mbfl_encoding **elist, unsigned int elist_size, bool strict)
+{
+	if (elist_size == 0) {
+		return NULL;
+	}
+	if (elist_size == 1) {
+		if (strict) {
+			return php_mb_check_encoding((const char*)in, in_len, *elist) ? *elist : NULL;
+		} else {
+			return *elist;
+		}
+	}
+	if (in_len == 0) {
+		return *elist;
+	}
+
+	uint32_t wchar_buf[128];
+	struct conversion_data {
+		const mbfl_encoding *enc;
+		unsigned char *in;
+		size_t in_len;
+		uint64_t demerits; /* Wide bit size to prevent overflow */
+		unsigned int state;
+	};
+	/* Allocate on stack; when we return, this array is automatically freed */
+	struct conversion_data *data = alloca(elist_size * sizeof(struct conversion_data));
+
+	for (int i = 0; i < elist_size; i++) {
+		data[i].enc = elist[i];
+		data[i].in = in;
+		data[i].in_len = in_len;
+		data[i].state = 0;
+		data[i].demerits = 0;
+	}
+
+	int finished = 0; /* For how many candidate encodings have we processed all the input? */
+	while (elist_size > 1 && finished < elist_size) {
```

Comment:
Nit: signed to unsigned conversion
```suggestion
	unsigned int finished = 0; /* For how many candidate encodings have we processed all the input? */
	while (elist_size > 1 && finished < elist_size) {
```

Reply:
Thanks, done.